### PR TITLE
S3 access grants plugin autoconfiguration

### PR DESCRIPTION
**Pull request by MatejNedic**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :pencil: Checklist

- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: docs/src/main/asciidoc/s3.adoc [hunk 1]
```diff
@@ -186,6 +186,22 @@ public class SimpleResourceLoadingBean {
...
+Since IAM polices and S3 Policies only support 10kbs size, S3 Access Grant is solving this by allowing fine grained access control over content in bucket.
+
+To use S3 Access Grants out of the box with `S3Client` and `S3Template` introduce following plugin:
+
+[source,xml]
+----
+<dependency>
+	<groupId>software.amazon.s3.accessgrants</groupId>
+	<artifactId>aws-s3-accessgrants-java-plugin</artifactId>
+	<optional>true</optional>
```
Comment (maciejwalkowiak) [inline docs/src/main/asciidoc/s3.adoc:201]: Here should be without optional, right?

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfigurationTests.java [hunk 2]
```diff
@@ -69,6 +71,30 @@ class S3AutoConfigurationTests {
...
 					CredentialsProviderAutoConfiguration.class, S3AutoConfiguration.class));
 
+	private final ApplicationContextRunner contextRunnerWithoutGrant = new ApplicationContextRunner()
+			.withPropertyValues("spring.cloud.aws.region.static:eu-west-1")
+			.withConfiguration(AutoConfigurations.of(AwsAutoConfiguration.class, RegionProviderAutoConfiguration.class,
+					CredentialsProviderAutoConfiguration.class, S3AutoConfiguration.class))
+			.withClassLoader(new FilteredClassLoader(S3AccessGrantsPlugin.class));
+
+	@Test
+	void testThatS3AccessGrantIdentityProviderIsSet() {
```
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfigurationTests.java:81]:
```suggestion
	void setsS3AccessGrantIdentityProvider() {
```

File: spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfigurationTests.java [hunk 3]
```diff
@@ -69,6 +71,30 @@ class S3AutoConfigurationTests {
...
+		contextRunner.run(context -> {
+			S3ClientBuilder builder = context.getBean(S3ClientBuilder.class);
+			ConfiguredAwsClient client = new ConfiguredAwsClient(builder.build());
+			assertThat(client.getIdentityProviders()).isInstanceOf(S3AccessGrantsIdentityProvider.class);
+		});
+	}
+
+	@Test
+	void testThatS3AccessGrantIdentityProviderIsNotSet() {
```
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfigurationTests.java:90]:
```suggestion
	void doesNotSetS3AccessGrantIdentityProvider() {
```

File: spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/properties/S3Properties.java [hunk 4]
```diff
@@ -93,6 +93,12 @@ public class S3Properties extends AwsClientProperties {
 	@NestedConfigurationProperty
 	private S3CrtClientProperties crt;
 
+	/**
+	 * If set to false if Access Grants does not find/return permissions, S3Client won't try to determine if policies
+	 * grant access If set to true fallback policies S3/IAM will be evaluated.
+	 */
+	private boolean enableFallback;
```
Comment (maciejwalkowiak) [inline spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/properties/S3Properties.java:100]: Perhaps this should be set on a nested property class specific to accesss grants plugin?

**Comment by MatejNedic** [comment]
@maciejwalkowiak all resolved

**Review by maciejwalkowiak — APPROVED** [review]

**Pull request merged**